PSReadLine/PlatformWindows.cs

```diff
@@ -1015,4 +1015,125 @@ private static void TerminateStragglers()
             }
         }
     }
+
+    private static readonly object _terminalOwnerThreadHandleLock = new();
+    private static IntPtr? _terminalOwnerThreadHandle;
+
+    /// <remarks>
+    /// <para>
+    ///     This method helps to find the thread id that owns the console window of the current process. The logic is
+    ///     that at the moment we are called, most likely a current foreground window owner will be the terminal window
+    ///     owner as well. We verify that by going up the process tree and if we find the foreground process owner among
+    ///     the current process' parent hierarchy, then we consider we found it.
+    /// </para>
+    /// <para>
+    ///     The terminal process is not always the direct parent of the current process, but may be higher in the
+    ///     process tree in case PowerShell is a child of some other console process.
+    /// </para>
+    /// <para>
+    ///     The result is cached for performance reasons. To avoid thread id reuse, we keep the thread handle as well.
```

Comment:
I guess we can add a check for `WaitForSingleObject(threadHandle, 0)` to figure out if a thread was invalidated.

This would still be open for a race condition, but only for a very little amount of time.

I wonder if we care for this now, actually.

In the worst case, if a thread owning the terminal for our process has been terminated, and we are still trying to process some key presses, we may try to access a wrong thread (if its id was reused already).

My vote is to still try to add as much protection as reasonably possible. But if you think we could just ignore this, then well we can.

Reply:
According to https://learn.microsoft.com/en-us/windows/win32/procthread/creating-windows-in-threads:
> The thread that creates the window owns the window and its associated message queue.

And according to https://learn.microsoft.com/en-us/windows/win32/procthread/terminating-a-thread:
> Terminating a thread has the following results:
>  * Any resources owned by the thread, such as windows and hooks, are freed.

So, in our case here, if the thread that created the terminal window gets terminated, then the terminal window itself is gone. Therefore, I think we can safely assume that the thread identified by the returned `tid` is alive as long as PowerShell is still alive.

Given the above, we should just cache the `tid` once we've verified it's the right one.

Reply:
I am still a little concerned, but alright, let's give this approach a try.